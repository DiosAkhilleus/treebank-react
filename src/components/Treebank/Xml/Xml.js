import React from 'react';

import TreebankContext from '../treebank-context';

import Xml from '../../Xml';

const XmlWithContext = () => (
  <TreebankContext.Consumer>
    {({ sentence, active, toggleActive }) => (
      <Xml sentence={sentence} active={active} toggleActive={toggleActive} />
    )}
  </TreebankContext.Consumer>
);

export default XmlWithContext;
