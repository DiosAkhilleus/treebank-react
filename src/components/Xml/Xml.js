import React, { Fragment } from 'react';

import { sentenceType } from '../../types';

import styles from './Xml.module.scss';

const attributesToJsx = (attributes) => (
  Object.entries(attributes).map(([key, value]) => (
    <Fragment key={key}>
      {' '}
      <span className={styles.attribute}>
        {key}
        =
      </span>
      <span className={styles.attributeText}>
        &quot;
        {value}
        &quot;
      </span>
    </Fragment>
  ))
);

const openingTag = (name, attributes, key, close = false, prefix = '') => (
  <Fragment key={key}>
    <span className={styles.bracket}>
      {prefix}
      &lt;
    </span>
    <span className={styles.element}>{name}</span>
    {attributesToJsx(attributes)}
    {close ? ' /' : ''}
    <span className={styles.bracket}>&gt;</span>
    <br />
  </Fragment>
);

const closingTag = (name, key) => (
  <Fragment key={key}>
    <span className={styles.bracket}>&lt;</span>
    <span className={styles.element}>
      /
      {name}
    </span>
    <span className={styles.bracket}>&gt;</span>
  </Fragment>
);

const Xml = ({ sentence }) => (
  <>
    {openingTag('sentence', sentence.$, 'sentence')}
    {sentence.word.map((word) => (
      openingTag('word', word.$, word.$.id, true, <>&nbsp;&nbsp;&nbsp;&nbsp;</>)
    ))}
    {closingTag('sentence', 'sentence-close')}
  </>
);

Xml.propTypes = {
  sentence: sentenceType.isRequired,
};

export default Xml;
