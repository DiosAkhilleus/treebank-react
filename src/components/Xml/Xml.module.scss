.xml {
  background-color: rgb(242, 242, 242);

  padding-top: 0.2rem;
  padding-bottom: 0.2rem;
  padding-left: 0.4rem;
  padding-right: 0.4rem;
}

.word {
  cursor: pointer;

  &:hover {
    background-color: #5BC8DC;
  }
}

.active {
  background-color: rgb(246, 217, 24);
}

.bracket {
  color: rgb(0, 0, 255);
}

.element {
  color: rgb(128, 0, 128);
}

.attribute {
  color: rgb(0, 0, 0);
}

.attributeText {
  color: rgb(255, 0, 0);
}
