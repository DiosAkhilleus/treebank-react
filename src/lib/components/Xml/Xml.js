import React, { Fragment } from 'react';
import { func } from 'prop-types';

import { sentenceType, wordType } from '../../types';

import styles from './Xml.module.scss';

const attributesToJsx = (attributes) => (
  Object.entries(attributes).map(([key, value]) => (
    <Fragment key={key}>
      {' '}
      <span className={styles.attribute}>
        {key}
        =
      </span>
      <span className={styles.attributeText}>
        &quot;
        {value}
        &quot;
      </span>
    </Fragment>
  ))
);

const openingTag = (name, attributes, key, close = false, prefix = '') => (
  <Fragment key={key}>
    <span className={styles.bracket}>
      {prefix}
      &lt;
    </span>
    <span className={styles.element}>{name}</span>
    {attributesToJsx(attributes)}
    {close ? ' /' : ''}
    <span className={styles.bracket}>&gt;</span>
    <br />
  </Fragment>
);

const closingTag = (name, key) => (
  <Fragment key={key}>
    <span className={styles.bracket}>&lt;</span>
    <span className={styles.element}>
      /
      {name}
    </span>
    <span className={styles.bracket}>&gt;</span>
  </Fragment>
);

const renderWord = (word, active, toggleActive) => {
  const { $: { id } } = word;
  const isActive = active && active.$.id === id;
  const className = isActive ? [styles.word, styles.active].join(' ') : styles.word;
  const onClick = () => {
    toggleActive(word);
  };
  const onKeyDown = (event) => {
    const { key } = event;

    if (key === 'Enter') {
      onClick(event);
    }
  };

  return (
    <div
      key={id}
      role="button"
      tabIndex="0"
      className={className}
      onClick={onClick}
      onKeyDown={onKeyDown}
    >
      {openingTag('word', word.$, id, true, <>&nbsp;&nbsp;&nbsp;&nbsp;</>)}
    </div>
  );
};

const Xml = ({ sentence, active, toggleActive }) => (
  <div className={styles.xml}>
    {openingTag('sentence', sentence.$, 'sentence')}
    {sentence.word.map((word) => (
      renderWord(word, active, toggleActive)
    ))}
    {closingTag('sentence', 'sentence-close')}
  </div>
);

Xml.propTypes = {
  sentence: sentenceType.isRequired,
  active: wordType,
  toggleActive: func,
};

Xml.defaultProps = {
  active: null,
  toggleActive: () => {},
};

export default Xml;
