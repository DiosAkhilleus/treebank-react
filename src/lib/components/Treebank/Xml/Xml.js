import React from 'react';

import SentenceContext from '../sentence-context';

import Xml from '../../Xml';

const XmlWithContext = () => (
  <SentenceContext.Consumer>
    {({
      sentence, active, toggleActive, highlight,
    }) => (
      <Xml
        sentence={sentence}
        active={active}
        toggleActive={toggleActive}
        highlight={highlight}
      />
    )}
  </SentenceContext.Consumer>
);

export default XmlWithContext;
