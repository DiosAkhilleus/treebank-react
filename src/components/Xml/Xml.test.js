import React from 'react';
import renderer from 'react-test-renderer';

import Xml from './Xml';

it('renders a sentence as XML', () => {
  const sentence = {
    $: {
      id: '1', document_id: 'test', subdoc: '2', span: '',
    },
    word: [
      {
        $: {
          id: '1', form: 'οὐκ', postag: 'd--------', relation: 'AuxZ', head: '1',
        },
      },
      {
        $: {
          id: '2', form: 'ᾔδη', lemma: 'οἶδα', postag: 'v1slia---', relation: 'PRED', head: '0',
        },
      },
    ],
  };

  const component = (
    <Xml sentence={sentence} />
  );
  const tree = renderer.create(component).toJSON();

  expect(tree).toMatchSnapshot();
});
