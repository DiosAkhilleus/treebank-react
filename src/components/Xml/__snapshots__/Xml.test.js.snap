// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a sentence as XML 1`] = `
Array [
  <span
    className="bracket"
  >
    
    &lt;
  </span>,
  <span
    className="element"
  >
    sentence
  </span>,
  " ",
  <span
    className="attribute"
  >
    id
    =
  </span>,
  <span
    className="attributeText"
  >
    "
    1
    "
  </span>,
  " ",
  <span
    className="attribute"
  >
    document_id
    =
  </span>,
  <span
    className="attributeText"
  >
    "
    test
    "
  </span>,
  " ",
  <span
    className="attribute"
  >
    subdoc
    =
  </span>,
  <span
    className="attributeText"
  >
    "
    2
    "
  </span>,
  " ",
  <span
    className="attribute"
  >
    span
    =
  </span>,
  <span
    className="attributeText"
  >
    "
    
    "
  </span>,
  "",
  <span
    className="bracket"
  >
    &gt;
  </span>,
  <br />,
  <span
    className="bracket"
  >
        
    &lt;
  </span>,
  <span
    className="element"
  >
    word
  </span>,
  " ",
  <span
    className="attribute"
  >
    id
    =
  </span>,
  <span
    className="attributeText"
  >
    "
    1
    "
  </span>,
  " ",
  <span
    className="attribute"
  >
    form
    =
  </span>,
  <span
    className="attributeText"
  >
    "
    οὐκ
    "
  </span>,
  " ",
  <span
    className="attribute"
  >
    postag
    =
  </span>,
  <span
    className="attributeText"
  >
    "
    d--------
    "
  </span>,
  " ",
  <span
    className="attribute"
  >
    relation
    =
  </span>,
  <span
    className="attributeText"
  >
    "
    AuxZ
    "
  </span>,
  " ",
  <span
    className="attribute"
  >
    head
    =
  </span>,
  <span
    className="attributeText"
  >
    "
    1
    "
  </span>,
  " /",
  <span
    className="bracket"
  >
    &gt;
  </span>,
  <br />,
  <span
    className="bracket"
  >
        
    &lt;
  </span>,
  <span
    className="element"
  >
    word
  </span>,
  " ",
  <span
    className="attribute"
  >
    id
    =
  </span>,
  <span
    className="attributeText"
  >
    "
    2
    "
  </span>,
  " ",
  <span
    className="attribute"
  >
    form
    =
  </span>,
  <span
    className="attributeText"
  >
    "
    ᾔδη
    "
  </span>,
  " ",
  <span
    className="attribute"
  >
    lemma
    =
  </span>,
  <span
    className="attributeText"
  >
    "
    οἶδα
    "
  </span>,
  " ",
  <span
    className="attribute"
  >
    postag
    =
  </span>,
  <span
    className="attributeText"
  >
    "
    v1slia---
    "
  </span>,
  " ",
  <span
    className="attribute"
  >
    relation
    =
  </span>,
  <span
    className="attributeText"
  >
    "
    PRED
    "
  </span>,
  " ",
  <span
    className="attribute"
  >
    head
    =
  </span>,
  <span
    className="attributeText"
  >
    "
    0
    "
  </span>,
  " /",
  <span
    className="bracket"
  >
    &gt;
  </span>,
  <br />,
  <span
    className="bracket"
  >
    &lt;
  </span>,
  <span
    className="element"
  >
    /
    sentence
  </span>,
  <span
    className="bracket"
  >
    &gt;
  </span>,
]
`;
